securesystemslib/signer/_signer.py

```diff
@@ -0,0 +1,264 @@
+# The argument is the name of the secret ("PIN", "passphrase", etc).
+# Return value is the secret string.
+SecretsHandler = Callable[[str], str]
+
+
+class Signer:
+    """Signer interface that supports multiple signing implementations.
+
+    Usage example:
+        signer = Signer.from_priv_key_uri("envvar:MYPRIVKEY", pub_key)
+        sig = signer.sign(b"data")
+
+    See SIGNER_FOR_URI_SCHEME for supported private key URI schemes. The
+    currently supported default schemes are:
+    * envvar: see SSlibSigner for details
+    * file: see SSlibSigner for details
+    """
+
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def sign(self, payload: bytes) -> Signature:
+        """Signs a given payload by the key assigned to the Signer instance.
+
+        Arguments:
+            payload: The bytes to be signed.
+
+        Returns:
+            Returns a "Signature" class instance.
+        """
+        raise NotImplementedError  # pragma: no cover
+
+    @classmethod
+    @abc.abstractmethod
+    def from_priv_key_uri(
+        cls,
+        priv_key_uri: str,
+        public_key: Key,
+        secrets_handler: Optional[SecretsHandler] = None,
+    ) -> "Signer":
+        """Factory constructor for a given private key URI
+
+        Returns a specific Signer instance based on the private key URI and the
+        supported uri schemes listed in SIGNER_FOR_URI_SCHEME.
+
+        Args:
+            priv_key_uri: URI that identifies the private key
+            public_key: Key that is the public portion of this private key
+            secrets_handler: Optional function that may be called if the
+                signer needs additional secrets (like a PIN or passphrase).
+                secrets_handler should return the requested secret string.
+
+        Raises:
+            ValueError: Incorrect arguments
+            Other Signer-specific errors: These could include OSErrors for
+                reading files or network errors for connecting to a KMS.
+        """
+
+        scheme, _, _ = priv_key_uri.partition(":")
+        if scheme not in SIGNER_FOR_URI_SCHEME:
+            raise ValueError(f"Unsupported private key scheme {scheme}")
+
+        signer = SIGNER_FOR_URI_SCHEME[scheme]
+        return signer.from_priv_key_uri(
+            priv_key_uri, public_key, secrets_handler
+        )
+
+
+class SSlibSigner(Signer):
+    """A securesystemslib signer implementation.
+
+    Provides a sign method to generate a cryptographic signature with a
+    securesystemslib-style rsa, ed25519 or ecdsa key. See keys module
+    for the supported types, schemes and hash algorithms.
+
+    SSlibSigners should be instantiated with Signer.from_priv_key_uri().
+    These private key URI schemes are supported:
+    * "envvar:<VAR>":
+        VAR is an environment variable with unencrypted private key content.
+           envvar:MYPRIVKEY
+    * "file:<PATH>?encrypted=[true|false]":
```

Comment:
Looking at the code, `true` can be anything but `false`. Is this true-ish behavior intended? I guess it makes sense.

Reply:
I figured it's the safest option to avoid somehow ending up with unencrypted passphrases if
* we later implement key creation and 
* you use "encrypted=yes" or "encrypted=True" or something